### DISCUSSION TITLE
How can I replace the Image used in Helm Chart to make it support ARM architecture?

### DISCUSSION THREAD

**Discussion by BloodBladeW** [mastodon/mastodon#22319] [Q&A] [answered]
The Bitmani image used in Helm Chart is not supported in the ARM architecture, and someone in [this issue](https://github.com/mastodon/chart/issues/22) recommended replacing it with [ZCube/bitnami-compat](https://github.com/ZCube/bitnami-compat), but I just recently got into using Kubernetes, so I don't really know how to do that yet, can anyone tell me what to edit?

**Comment by magsol** [comment, marked as answer]
**Some asides**

I'm working on a series of blog posts on the topic so I'll hopefully have more detail for you soon, but briefly, here are the files in this repo you'll need to modify:

 - `values.yaml`. This is where all the specific helm chart configuration entries are, and you'll need to change at least a few of them. You _could_ instead specify the values as part of the `helm install` command, but this is a bit more complicated for numerous reasons, so instead I'd recommend just modifying this yaml file.
 - `Chart.yaml`. I'd say that "normally" you would NOT need to modify this file, but because of a weird quirk of the various dependencies, you'll need to add an entry to this.

Given some recent merged PRs, you shouldn't have to modify any of the template files (thank goodness).

One final aside: the Mastodon chart (or, more specifically, the Mastodon _image_) actually DOES support ARM. The reason it doesn't work out-of-the-box with ARM is because the Mastodon chart relies on redis, postgres, and elasticsearch, and the images that are pulled for each of these are the ones maintained by bitnami, and bitnami does NOT support ARM.

Hence, to get the full chart to work on ARM, you have to get these three dependencies to work on ARM. (I know there are some dev discussions around decoupling the Mastodon chart from the dependencies, but for now and the near future that's not an option)

**Dependencies on ARM**

As stated, to get the full chart to work on ARM, you have to install the three dependencies on ARM. This amounts to three steps:
 - DISABLE the dependencies in the Mastodon chart
 - ADD a common library to Chart.yaml
 - install the dependencies SEPARATELY

I'll take these one at a time.

**1: Disable the dependencies in the Mastodon Helm chart**

This is probably the easiest step. Look for the top-level entries in `values.yaml` for `elasticsearch:`, `postgresql:`, and `redis:`. Under each of them should be an `enabled: true` directive. Simply change the `true` to `false`. Congrats! All three are disabled! By simply doing this, your Mastodon chart should now work on ARM. Mastodon _itself_ will be broken without these dependencies to connect to, but the Chart itself is now fully ARM compatible :)

You do need to do a few housekeeping tasks after disabling some of these dependencies, so that Mastodon will know where to find the _separate_ instances you spin up later. For example, [here are the fields for postgres you need to fill out](https://github.com/mastodon/chart#postgresql-passwords) (yes, even though you disabled it!). Same with redis (you'll need to create a `Secret` under `redis.auth.existingSecret`). Nothing extra for elasticsearch.

**2: Add a common library to Chart.yaml**

This is easy, just annoying. The reason this is needed is that, by disabling all the dependencies, ALL the bitnami charts have been removed from the Mastodon chart, including what's called a "common library" chart by bitnami that, weirdly, still seems to be needed. All you need to do is add the following yaml underneath the `dependencies:` section of `Chart.yaml`:

```yaml
 - name: common
   version: 1.0.0
   repository: https://charts.bitnami.com/bitnami
```

Leave the other entries as-is; all you have to do is add this new one to the bottom.

**3: Install the dependencies separately**

Now the hardest part: you'll need to install and configure each of redis, postgres, and elasticsearch separately. This took some doing on my end to find the configurations that worked--I would also consider myself a beginner so a lot of this I learned from trial and error, a bunch of google searching, and from endlessly bothering the Mastodon devs (thank you all again for your patience!).

I ended up just installing each via their own Helm charts, specifically the bitnami charts. BUT WAIT you're asking, I thought you said the bitnami charts weren't supported by ARM? The bitnami _images_ don't support ARM, but: the bitnami Helm charts have configuration options whereby you can essentially slot in the _official_ images (built by redis, postgres, and elasticsearch respectively) that _do_ support ARM, while still using the rest of bitnami's charts. Doing it this way ensures maximum configuration overlap between the dependencies once you've installed them and with Mastodon.

For example, here's the full command I used to install postgres (what I'd argue was the easiest dependency to install separately):

```bash

helm install postgres bitnami/postgresql \
  --set global.postgresql.auth.postgresPassword="should_match_whatevers_in_mastodon_values_yaml_for_postgres_password" \
  --set global.postgresql.auth.username="should_match_whatevers_in_mastodon_values_yaml_for_username" \
  --set global.postgresql.auth.password="should_match_whatevers_in_mastodon_values_yaml_for_password" \
  --set global.postgresql.auth.database="should_match_whatevers_in_mastodon_values_yaml_for_database" \
  --set image.repository="postgres" \
  --set image.tag="15.0"
```

What's happening here: the first line runs the helm install (specifying `bitnami/postgresql` means I'm still using bitnami's helm chart for the install); the 2nd-4th lines all set up postgres database configuration _that should match exactly corresponding entries in the Mastodon `values.yaml` section for postgres_; the 5th line is the magic that allow for using the _official_ postgres images (which _are_ ARM-compatible) rather than the custom-built bitnami ones (which are _not_ ARM-compatible); the 6th is just so that we make sure to use the same version of postgres that Mastodon would've used otherwise.

Rinse and repeat for redis, though the wrinkle here making sure it could find the `Secret` you created earlier with the redis password.

```bash

helm install redis bitnami/redis \
  --set global.redis.password="should_match_whatevers_in_mastodon_values_yaml_for_redis_password" \
  --set architecture="standalone" \
  --set image.repository="redis" \
  --set image.tag="7.0"
```

Again, the first line runs helm install on the bitnami chart for redis; the 2nd line contains the password you set in the Mastodon helm chart and which you stored in the `Secret`; the 3rd line was to minimize replication (since I'm running on Raspberry Pi cluster I don't have a lot of extra resources); the 4th line is, once again, the magic to pull the _official_ redis image that will run on ARM, and the 5th is again to use the same version of redis as that which Mastodon uses.

Elasticsearch... requires a whole post unto itself. Short version: I got the official version to work, but it's _extremely_ resource-intensive, so much so that I had to drop elasticsearch entirely; my Raspberry Pi cluster just didn't have the memory/CPU to run it, no matter how much I stripped it down in the configuration. At the suggestion of some of the Mastodon devs, I instead used [OpenSearch](https://github.com/opensearch-project), which some folks have already posted in issues as a potential elasticsearch alternative. It took a little configuration to get it working; not too bad, but had to be done _precisely_ or it wouldn't work at all. I'll post about the details of this one later since this has already gone on for quite awhile, and this isn't even required to get Mastodon working (just a "nice to have").

**4: Install Mastodon!**

Yes, at this point you should be ready to install Mastodon itself! Per the official instructions, this comprises two commands:

```bash
helm dep update

helm install mastodon ./ -f values.yaml
```

The first command runs a dependency update; this is _crucial_, as it will download that "common" library we added earlier to `Chart.yaml`. The second command (assuming you run it from within the [`chart/` directory](https://github.com/mastodon/chart)) will run the Mastodon Helm chart and plug in the configuration values you entered into `values.yaml`. If all goes well, your instance should be up and running after a few moments!

**Questions? Concerns? Complaints? Criticisms?** Like I said I hope to have a few blog posts coming out soon. But in the meantime you're welcome to post your questions, especially if something wasn't clear.

**Reply by BloodBladeW** [reply]
Thank you very much, followed your method and ran Mastodon successfully.
I have more memory so I tried installing ElasticSearch as per Elastic's website, but Mastodon doesn't accept its certificate and I don't know how to get them to use http to connect to each other.
Maybe I should try to use OpenSearch too, is there anything I should be aware of when using it?

**Reply by magsol** [reply]
If you want to go the elasticsearch route, you can still pull the same trick--using the bitnami helm chart with the elasticsearch image--to try and keep things as compatible as possible. But it takes a considerable amount of additional configuration, in no small part because for some reason bitnami's elasticsearch helm chart [uses a second "helper" image](https://hub.docker.com/r/bitnami/bitnami-shell/) to get things off the ground. This "helper" image _also_ needs to be swapped for [one that is ARM-compatible](https://github.com/zcube/bitnami-compat/pkgs/container/bitnami-compat%2Fbitnami-shell).

I have a working helm configuration if you want to try this! But if you'd rather not deal with all that, OpenSearch is probably on the whole simpler. There are just some very specific changes that need to be made [to its own values.yaml](https://github.com/opensearch-project/helm-charts/blob/main/charts/opensearch/values.yaml).
 - between lines 61 and 62 (underneath `security:`), add: `disabled: true`
 - on line 297, change `securityConfig.enabled` to `false`

These changes should actually obviate the need for certificates :)

I made a few other changes too but these were specific to the low-resource environment it was operating in (setting replicas to 1, decreasing the JavaOpts heap sizes, etc) and so you probably shouldn't need them.

**Reply by BloodBladeW** [reply]
I tried to search in Bitnami's ElasticSearch Chart using the keyword "shell" and found and modified two image repositories in values.yaml.
Shockingly, when running them, it used up all the remaining memory (18GB + 7GB Swap)!
I then deleted it and used OpenSearch the way you did, and although I ran into a [few hiccups](https://github.com/mastodon/mastodon/issues/18625), it eventually ran successfully and only used 3GB of memory. Now, all parts are running successfully.
Thanks again for your help.

**Reply by magsol** [reply]
Ah yes, the progress bar caused me some headaches too. Disabling it fixed everything 👍 Glad you got it all working!
